airavata-api/src/main/java/org/apache/airavata/api/server/handler/AiravataServerHandler.java

```diff
@@ -2124,64 +1651,23 @@ public List<OutputDataObjectType> getIntermediateOutputs(AuthzToken authzToken,
-            }
-
-            // Figure out if there are any currently running intermediate output fetching processes for outputNames
-            // First, find any existing intermediate output fetch processes for outputNames
-            List<ProcessModel> intermediateOutputFetchProcesses = existingExperiment.getProcesses().stream()
-                    .filter(p -> {
-                        // Filter out completed or failed processes
-                        if (p.getProcessStatusesSize() > 0) {
-                            ProcessStatus latestStatus = p.getProcessStatuses().get(p.getProcessStatusesSize() - 1);
-                            if (latestStatus.getState() == ProcessState.COMPLETED
-                                    || latestStatus.getState() == ProcessState.FAILED) {
-                                return false;
-                            }
-                        }
-                        return true;
-                    })
-                    .filter(p -> p.getTasks().stream().allMatch(t -> t.getTaskType() == TaskTypes.OUTPUT_FETCHING))
-                    .filter(p -> p.getProcessOutputs().stream().anyMatch(o -> outputNames.contains(o.getName())))
-                    .collect(Collectors.toList());
-            if (!intermediateOutputFetchProcesses.isEmpty()) {
-                throw new InvalidRequestException(
-                        "There are already intermediate output fetching tasks running for those outputs.");
-            }
-
-            String gatewayId = authzToken.getClaimsMap().get(Constants.GATEWAY_ID);
-            submitExperimentIntermediateOutputsEvent(gatewayId, airavataExperimentId, outputNames);
-            registryClientPool.returnResource(regClient);
-            sharingClientPool.returnResource(sharingClient);
-        } catch (InvalidRequestException | AuthorizationException e) {
-            logger.error(e.getMessage(), e);
-            registryClientPool.returnResource(regClient);
+            airavataService.validateAndFetchIntermediateOutputs(
+                    sharingClient, authzToken, airavataExperimentId, outputNames, experimentPublisher);
             sharingClientPool.returnResource(sharingClient);
-            throw e;
         } catch (Exception e) {
+            if (e.getMessage() != null && e.getMessage().contains("does not have WRITE access")) {
+                logger.error(e.getMessage(), e);
+                sharingClientPool.returnResource(sharingClient);
+                throw new AuthorizationException(e.getMessage());
+            } else if (e.getMessage() != null
+                    && (e.getMessage().contains("does not have currently ACTIVE job")
+                            || e.getMessage().contains("already intermediate output fetching tasks"))) {
+                logger.error(e.getMessage(), e);
+                sharingClientPool.returnResource(sharingClient);
+                throw new InvalidRequestException(e.getMessage());
```

Comment:
It's always good to avoid these string checks. Handling different exception types (before code change) would be better